Inferences: add notifications, run automatically in background

noerw opened the pull request:
fixes #66 
fixes #11 

this is an extract of `feature/improve-inferences`, without the commits for #72 , in order to streamline review & merge.

schrooom committed:
#67: refactored inference-namespace

schrooom committed:
#67: run inferences with location-update (~every 4 hours)

schrooom committed:
#67: refactor trajectory/sqlite-service

felixerdy committed:
intoduce noInferencesFound result

felixerdy committed:
show notification on new inferenceResult

schrooom committed:
#67: minor adjustment to trajectory-card label

schrooom committed:
#66 - simplified inference exec

noerw commented on src/app/shared-services/inferences/inference.service.ts:
```diff
@@ -0,0 +1,88 @@
+
+    if (inference.status === InferenceResultStatus.successful) {
+      this.localNotifications.schedule({
+        id: Math.random() * 1000000,
```
This is a dirty hack, we should avoid it if possible.
On android, notifications work without providing `id` explicitly (unless we want to edit notifications after posting them, which we don't). Need to check on iOS if that works there as well?

noerw commented on src/app/shared-services/trajectory.service.ts:
```diff
@@ -94,7 +94,6 @@ export class TrajectoryService {
   }
 
   deleteTrajectory(t: TrajectoryMeta) {
-    this.locationService.stop()
```
It's weird that this was here in the first place, but are you sure that we can safely remove it?

schrooom replied:
I did not come across unexpected side-effects from this. Most importantly leaving this as it is creates a cyclic dependency:

→ `LocationService` triggers infering using `InferenceService`
→ `InferenceService` loads trajectories using `TrajectoryService`
→ `TrajectoryService` stops tracking using `LocationService` (therefore this one was deleted)

I'll simply add `this.locationService.stop()` to the event of the delete-button.

noerw commented on src/app/shared-services/trajectory.service.ts:
```diff
@@ -50,6 +45,11 @@ export class TrajectoryService {
   }
 
+  getFullUserTrack(): Observable<Trajectory> {
+    if (!this.db.isSupported()) return from(Promise.resolve(undefined))
```
This would be better handled at the layer of the sqlite service.
I think in that layer we currently never resolve when DB is not supported. If that doesn't work for `getFullUserTrack()`, better change that to resolve to undefined, instead of adding more logic here?
But leave it as is, if that requires a larger refactor

schrooom replied:
Actually we did do the same in line 38. Nevertheless you are right and whether the DB is supported or not is already checked in the `SQLiteService`.

noerw commented on src/app/shared-services/inferences/inference.service.ts:
```diff
@@ -0,0 +1,88 @@
+    return diffInMinutes < InferenceService.inferenceIntervalMinutes
+  }
+
+  loadPersistedInferences(trajectoryId: string): InferenceResult {
```
This will be async once implemented, we could define its callsignature as such already

noerw commented on src/app/shared-services/location.service.ts:
```diff
@@ -109,13 +109,15 @@ export class LocationService implements OnDestroy {
           speed,
         })
 
+        this.inf.generateUserInference()
```
This is executed asynchronously, right? In that case, please add a comment noting that.

But are you sure we can run it asynchronously? [According to background location plugin docs](https://github.com/mauron85/cordova-plugin-background-geolocation#example), execution time is limited in this callback for iOS, and should be wrapped in a "transaction" like this:
```js
  BackgroundGeolocation.on('location', function(location) {
    // handle your locations here
    // to perform long running operation on iOS
    // you need to create background task
    BackgroundGeolocation.startTask(function(taskKey) {
      // execute long running task
      // eg. ajax post location
      // IMPORTANT: task has to be ended by endTask
      BackgroundGeolocation.endTask(taskKey);
    });
  });
```

schrooom replied:
I'll need to take a closer look at this.

Potentially problematic: This probably triggers the usual background processing in iOS, which is pretty strictly controlled by the OS. In the use-case of the learning app, which may not to be actively used very much, the OS could rate our app as low-priority and thus give it very little to no background time for such tasks. The location updates on the other hand are triggered quite regularly, therefore the execution is currently done directly there. Will update on this asap.

noerw commented:
@schrooom, @felixerdy I did a quick review :)

noerw commented:
Also, test fail is related to the changes:

```
Chrome Headless 90.0.4430.93 (Linux x86_64) SelectTrajectoryPage should create FAILED
	Failed: R3InjectorError(DynamicTestModule)[LocationService -> InferenceService -> TrajectoryService -> TrajectoryService]: 
	  NullInjectorError: No provider for TrajectoryService!
	error properties: Object({ ngTempTokenPath: null, ngTokenPath: [ 'LocationService', 'InferenceService', 'TrajectoryService', 'TrajectoryService' ] })
```

schrooom committed:
- #74: Fixes review comments

- remove id from notifications
- re-add tracking-stop, when deleting user-trajectory
- change signature of loadPersistedInferences(..) to async
- remove redundant DB check

schrooom committed:
- #74: Fixes review comments II

- fixed a few tests
- initial value for inferences

schrooom committed:
#74: Refactor - moved trajectory-services

schrooom committed:
- #74: Fixes review comments III

 - generate inferences in BackgroundLocation-Task triggerd by location updates
 - introducing NotificationService for creating notifications
   - re-add IDs for notifications

schrooom committed:
#74: refactored inference types.ts

schrooom commented:
I've fixed/reworked the mentioned aspects :)

noerw commented:
I'd send an approving review, but can't as I created this PR myself ;)
so: LGTM
